### ISSUE TITLE
Form Submission Error Display

### ISSUE THREAD

**Issue by scottluptowski** [final-form/final-form#4]
Hi, thanks so much for this library! Looks really nice and minimal.

I'm designing an interaction in which form errors are shown to the user after they attempt to submit a form.

If there are errors, I want the errors to stay on the screen until the next time the user attempts the submission. In other words, I do not want an error to appear/disappear as a user types the text that swaps it from invalid to valid.

This is not currently possible with Final Form as I understand it. If a form field is `invalid` at the point of one submission,  making it `valid` will wipe away its former `invalid`ness, removing the error display. There is no way to model "this field was previously submitted, and was invalid at that time" for display purposes. `submitFailed` is not `true`, because the submit action never happened due to the presence of errors.

Is there a way to solve this problem with Final Form today?

Thanks again!

**Comment by erikras** [comment]
Fantastic reporting! Hmm... I faced this with `redux-form`. I think perhaps the answer is that `submitError` should be cleared for each field on `change`. Does that sound like it might fix the problem? (and form-wide `FORM_ERROR` should be cleared on the change of _any_ field).

**Comment by erikras** [comment]
Well. I also recall that people complained about this when I did it in `redux-form`, which resulted in a `persistentSubmitErrors` config parameter.

**Comment by jedrichards** [comment]
I always found the flags and state that `redux-form` used to manage the meta state of the form across possibly complex life cycles (submit > submit error > edit > submit again) hard to understand, and a bit clunky. While the state of the form as derived from the state of each individual field felt rock solid, the ephemeral state of the overall form lifecycle through multiple submissions felt less so.

Is there a way to push this sort of thing into user land a bit more, perhaps by modelling the state differently or more explicitly, rather than adding config parameters?

I haven't thought about it much, but could historical errors and state be modelled somehow on fields? Then the user could keep track of submissions, and conditionally display the last error on a field if needed?

tldr; Feels like the ideal solution would be that display of "persistent submission errors" and the like should be able to be _derived from the state_, rather than imperatively tweaking how state gets mutated within the lib?

**Comment by erikras** [comment]
Excellent comment, @jedrichards. This is why, in Final Form, I specifically broke out the `submitError` into its own value, separate from validation errors. Would a `dirtySinceLastSubmit` flag help? Might it help to outline all possible scenarios?

**Comment by jedrichards** [comment]
`dirtySinceLastCommit` as a form wide state value? Yeah that could be useful ...

So users could do something like the below to work out whether to show the submit error?

```
const showSubmitError = submitError && !dirtySinceLastSubmit
```

**Comment by erikras** [comment]
`dirtySinceLastCommit` could be both form-wide and field-specific.

**Comment by klis87** [comment]
@erikras I think this would be extremely helpful and flexible solution for server side errors and with this flag everyone could do whatever desired, for both field and form level errors.

Is there any workaround for now to achieve sth like this?

**Comment by klis87** [comment]
Also, apart from dirtySinceLastSubmit flag, what do you think about additionally adding a new action in form render props, like setSubmitErrors or clearSubmitError?

**Comment by erikras** [comment]
Published `dirtySinceLastCommit` fix in [`v3.1.0`](https://github.com/final-form/final-form/releases/tag/v3.1.0).
